File: hackerrank/contests/algorithms/infinitum10/degree_of_an_algebraic_number.cpp
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <algorithm>
#include <math.h>


typedef long long ll;

using namespace std;

int A[101];

inline
void solve() {
  int n;
  scanf("%d", &n);
  for (int i = 0; i < n; ++ i) {
    scanf("%d", A+i);
  }

  for (int i = 0; i < n; ++ i) {
    int bd = (int) sqrt(A[i]) + 1;
    for (int a = bd; a > 1; -- a) {
      if (A[i] % (a*a) == 0) {
        A[i] /= a*a;
        break;
      }
    }
  }

  sort(A, A+n);
  ll cnt = unique(A, A+n) - A;
  if (A[0] == 1) {
    cnt -= 1;
  }
  printf("%lld\n", 1ll<<cnt);

}

int main() {
  int t;
  scanf("%d", &t);
  for (;t--;) {
    solve();
  }

  return 0;
}

File: hackerrank/contests/algorithms/infinitum10/degree_of_an_algebraic_number.py
# -*- coding: utf-8 -*-

import math

t = int(raw_input())

def solve():
  n = int(raw_input())
  A = map( int, raw_input().split() )
  A = set(A)

  cnt = 0
  while 1:
    R = []
    for a in A:
      if int(math.sqrt(a)) ** 2 == a:
        R.append(a)

    for r in R:
      A.remove(r)

    if not A:
      break

    cnt += 1

    N = set()
    for i in A:
      for j in A:
        if i != j:
          N.add(i*j)

    A = N

  print 2**cnt

for _ in xrange(t):
  solve()
